### PR TITLE
Add CryptoMB private key provider to contrib.

### PR DESCRIPTION
**Pull request by ipuustin**
Commit Message:

Intel's IPP (Integrated Performance Primitives) crypto library has support for multi-buffer crypto operations. Briefly, multi-buffer
cryptography is implemented with AVX-512 instructions using a SIMD (single instruction, multiple data) mechanism. Up to eight RSA or ECDSA operations are gathered together into a buffer and processed at the same time, providing potentially improved performance. The AVX-512 instructions are available on recently launched 3rd generation Xeon Scalable server processors (Ice Lake server) processors.

This commit adds a private key provider to accelerate RSA and ECDSA crypto operations on recent Intel Xeon processors. Every worker thread has a queue of up-to-eight crypto operations. When the queue is full or when the timer is triggered, the queue is processed and all the pending handshakes are notified.

The potential performance benefit depends on many factors: the size of the cpuset Envoy is running on, incoming traffic pattern, encryption type (RSA or ECDSA), and key size. In my own testing I saw the biggest performance increase when long RSA keys were used on an Envoy running in a fairly limited environment serving lots of new incoming TLS requests. For more details, see this Intel whitepaper which contains some more information about the AVX-512 instructions and potential performance increase: https://www.intel.com/content/www/us/en/architecture-and-technology/crypto-acceleration-in-xeon-scalable-processors-wp.html

Additional Description:

One new dependency is introduced: Intel’s ipp-crypto library. Currently the PR is using a development version of ipp-crypto because BoringSSL support is not yet part of any release. The ipp-crypto team has indicated that BoringSSL version will be included in future ipp-crypto releases.

Basic tests are provided, and a fake library interface is included for testing on systems without the required AVX-512 instruction set.

Risk Level: Medium (TLS security feature, not enabled by default)
Testing: Unit tests
Docs Changes: API interface is documented
Release Notes: Added CryptoMB private key provider to contrib.
Platform Specific Features: Requires Intel 3rd generation Xeon Scalable server processor for the AVX-512 IFMA instruction set.
Fixes: #15871

### CODE REVIEW

File: contrib/cryptomb/private_key_providers/test/test_data/ecdsa-p256.pem
Comment (ipuustin) [reply contrib/cryptomb/private_key_providers/test/test_data/ecdsa-p256.pem]: I added a script and regenerated the keys with it.

File: contrib/cryptomb/private_key_providers/source/ipp_crypto_disabled_impl.h
Comment (ipuustin) [reply contrib/cryptomb/private_key_providers/source/ipp_crypto_disabled_impl.h]: Ok, done. Waiting to see if ARM build and tests pass in CI...

File: tools/extensions/extensions_check.py
Comment (ipuustin) [reply tools/extensions/extensions_check.py:52]: Turns out it's needed for contrib extensions too. Reverting my change causes this error message:

```
ExtensionsChecker NOTICE [metadata] Running check
ExtensionsChecker ERROR [metadata] Unknown extension category for envoy.tls.key_providers.cryptomb: envoy.tls.key_providers. Please add it to tools/extensions/extensions_check.py
ExtensionsChecker ERROR [metadata] Check failed
ExtensionsChecker ERROR Summary
--------------------------------------------------------------------------------
ExtensionsChecker metadata:
Unknown extension category for envoy.tls.key_providers.cryptomb: envoy.tls.key_providers. Please add it to tools/extensions/extensions_check.py

ExtensionsChecker ERROR {'success': 0, 'errors': 1, 'warnings': 0, 'failed': {'metadata': 1}, 'warned': {}, 'succeeded': {}}
```

File: contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc
Comment (ipuustin) [inline contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc:14]: @agl @davidben Hi Adam and David! Do you think it would be possible to expose `ec_random_nonzero_scalar()`, `ec_scalar_to_bytes()` and the related `EC_SCALAR` data type in BoringSSL public API? They are required if you want to calculate ECDSA ephemeral keys the way BoringSSL does it.

Comment (agl) [reply contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc:14]: I don't think we're likely to do that because we keep eyeing whether we can put a version-skew coupling around the FIPS module.

(And it goes without saying that copy-pasting bits of internal headers is invalid and this change cannot land while doing that.)

**Review by agl — CHANGES_REQUESTED** [review]
I think this is the wrong direction for Envoy.

The feedback that we get is that TLS handshaking isn't a major problem and isn't worth significant investments. If one did want to invest there, however, pushing forward [batching](https://datatracker.ietf.org/doc/html/draft-davidben-tls-batch-signing-02) of TLS handshakes has dramatically better payoff and makes far more sense.

File: contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc
Comment (ipuustin) [reply contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc:14]: In this case, do you think there could be some other way how BoringSSL could provide access to this functionality? It seems to me that the only non-trivial function needed for this is `bn_rand_range_words()` and the functions it calls. I would love to use ready-made functions rather than implementing it myself. OpenSSL has library access to ECDSA ephemeral key generation, and I think I can't be the only person needing this.

**Comment by ipuustin** [comment]
> I think this is the wrong direction for Envoy.
> 
> The feedback that we get is that TLS handshaking isn't a major problem and isn't worth significant investments. If one did want to invest there, however, pushing forward [batching](https://datatracker.ietf.org/doc/html/draft-davidben-tls-batch-signing-02) of TLS handshakes has dramatically better payoff and makes far more sense.

I think we have been getting different feedback about this -- I'm hearing that TLS handshake cost is a real problem for edge use cases. Also many people already have processors with these AVX-512 instructions, so just turning the feature on and seeing if there are performance gains will probably not be a "significant investment" for them. If you are interested in the performance numbers (the expected payoff), [this whitepaper](https://www.intel.com/content/www/us/en/architecture-and-technology/crypto-acceleration-in-xeon-scalable-processors-wp.html) has some public numbers measured with HAProxy. Of course, you can try out the PR yourself and see how much performance gains you see with your setup.

Also, please note that this extension is an optional feature of Envoy, available to people who want to use it. It in no way excludes other possible TLS acceleration mechanisms.

**Comment by ggreenway** [comment]
> I think this is the wrong direction for Envoy.
> 
> The feedback that we get is that TLS handshaking isn't a major problem and isn't worth significant investments. If one did want to invest there, however, pushing forward [batching](https://datatracker.ietf.org/doc/html/draft-davidben-tls-batch-signing-02) of TLS handshakes has dramatically better payoff and makes far more sense.

@agl Thanks for the feedback on this. This extension is going into Envoy's `contrib/` tree, which is not included by default in builds and contains less scrutinized, less reviewed, more experimental extensions. Given that, I think it's fine to add this (although it's also very much fine for BoringSSL to not wish to add public APIs to support it), and see if people use it and get a speed boost.

**Comment by davidben** [comment]
You all definitely should not merge this while using BoringSSL internal APIs. Those are internal implementation details of the EC library. Using them _will_ break your build or, worse, accidentally get the ABI wrong and introduce a memory error.

File: contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc
Comment (ggreenway) [inline contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc:27]: I hadn't realized earlier that you're accessing an internal BoringSSL symbol here. This could break anytime we update the Envoy BoringSSL version. One option is that if this extension stops compiling, or tests stop passing, during a version update, we remove it from the contrib build until you can fix it. But I'm somewhat concerned about the increased friction of being able to handle core dependency updates from this.

cc @envoyproxy/senior-maintainers

**Review by ggreenway — CHANGES_REQUESTED** [review]

File: contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc
Comment (mattklein123) [reply contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc:27]: Yeah IMO we need to fix this before merging.

**Comment by ggreenway** [comment]
Note: In my above comment about `contrib`, I'm referring to whether or not it is a performance boost not gating merging this change; we'll try it. 

Accessing non-public BoringSSL APIs is a separate issue, that we do need to resolve.

**Comment by ipuustin** [comment]
> You all definitely should not merge this while using BoringSSL internal APIs and copy-and-pasting internal third-party headers. Those are internal implementation details of the EC library. Using them _will_ break your build or, worse, accidentally get the ABI wrong and introduce a memory error.

I agree that using the internal API is definitely a code smell, and that's why I asked your opinion if the symbols (or the BN function they call) could be made public, or some other way be found to calculate the value using BoringSSL public API. If that's not possible, I can think of three options:

1) I write the ephemeral key calculation according to https://tools.ietf.org/html/rfc6979#section-3.2. I'm just sort of reluctant writing crypto code of my own (for obvious reasons).
2) I copy the relevant BoringSSL implementation into the private key provider or a small helper lib. The problem of this approach is code duplication and that any improvements (including security fixes) to the code in BoringSSL need to be manually tracked.
3) I remove the ECDSA support from the private key provider (at least for now). ECDSA is faster than RSA and so the ECDSA performance benefits are not that big. The ECDSA signing could just be done using regular BoringSSL functions.

**Comment by ggreenway** [comment]
I think option 3 is a good starting point. After getting it working with RSA, if there's demand, we can revisit for ECDSA. I dislike both 1 and 2, for the reasons you've given.

**Comment by ipuustin** [comment]
> I think option 3 is a good starting point. After getting it working with RSA, if there's demand, we can revisit for ECDSA. I dislike both 1 and 2, for the reasons you've given.

Ok, I went for option 3 here.

File: api/contrib/envoy/extensions/private_key_providers/cryptomb/v3alpha/cryptomb.proto
Comment (ipuustin) [reply api/contrib/envoy/extensions/private_key_providers/cryptomb/v3alpha/cryptomb.proto]: That's a good idea, fixed.

File: api/contrib/envoy/extensions/private_key_providers/cryptomb/v3alpha/cryptomb.proto
Comment (ipuustin) [reply api/contrib/envoy/extensions/private_key_providers/cryptomb/v3alpha/cryptomb.proto]: Originally I had reserved the 0ms case for synchronous operation, needed for running the ssl_socket_test.cc tests. Since we moved to contrib we no longer have those tests, so I think requiring the delay to be positive duration makes sense. I fixed this now.

File: api/contrib/envoy/extensions/private_key_providers/cryptomb/v3alpha/cryptomb.proto
Comment (ggreenway) [inline api/contrib/envoy/extensions/private_key_providers/cryptomb/v3alpha/cryptomb.proto:23]: Update docs: this now does "normal" processing for ecdsa.

File: docs/root/intro/arch_overview/security/ssl.rst
Comment (ggreenway) [reply docs/root/intro/arch_overview/security/ssl.rst:225]: I think that's because those filters existed before we had `contrib`, so their docs are still in core docs. That should get cleaned up at some point. But new `contrib` extensions shouldn't add core docs.

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait

**Review by ggreenway — CHANGES_REQUESTED** [review]
You can move the cmake stuff into your extension like this: 

```
diff --git a/bazel/foreign_cc/BUILD b/bazel/foreign_cc/BUILD
index 199d9dd241..2c9b481282 100644
--- a/bazel/foreign_cc/BUILD
+++ b/bazel/foreign_cc/BUILD
@@ -236,21 +236,6 @@ envoy_cmake_external(
     }),
 )
 
-envoy_cmake_external(
-    name = "ipp-crypto",
-    cache_entries = {
-        "BORINGSSL": "on",
-    },
-    defines = [
-        "OPENSSL_USE_STATIC_LIBS=TRUE",
-    ],
-    lib_source = "@com_github_intel_ipp_crypto_crypto_mb//:all",
-    static_libraries = ["libcrypto_mb.a"],
-    tags = ["skip_on_windows"],
-    working_directory = "sources/ippcp/crypto_mb",
-    deps = ["@boringssl//:ssl"],
-)
-
 envoy_cmake_external(
     name = "llvm",
     cache_entries = {
diff --git a/contrib/cryptomb/private_key_providers/source/BUILD b/contrib/cryptomb/private_key_providers/source/BUILD
index 4288681993..00265c8de5 100644
--- a/contrib/cryptomb/private_key_providers/source/BUILD
+++ b/contrib/cryptomb/private_key_providers/source/BUILD
@@ -2,6 +2,7 @@ load(
     "//bazel:envoy_build_system.bzl",
     "envoy_cc_contrib_extension",
     "envoy_cc_library",
+    "envoy_cmake_external",
     "envoy_contrib_package",
 )
 
@@ -9,6 +10,21 @@ licenses(["notice"])  # Apache 2
 
 envoy_contrib_package()
 
+envoy_cmake_external(
+    name = "ipp-crypto",
+    cache_entries = {
+        "BORINGSSL": "on",
+    },
+    defines = [
+        "OPENSSL_USE_STATIC_LIBS=TRUE",
+    ],
+    lib_source = "@com_github_intel_ipp_crypto_crypto_mb//:all",
+    static_libraries = ["libcrypto_mb.a"],
+    tags = ["skip_on_windows"],
+    working_directory = "sources/ippcp/crypto_mb",
+    deps = ["@boringssl//:ssl"],
+)
+
 envoy_cc_library(
     name = "ipp_crypto_wrapper_lib",
     hdrs = ["ipp_crypto.h"] + select({
@@ -28,7 +44,7 @@ envoy_cc_library(
     repository = "@envoy",
     deps = select({
         "//bazel:linux_x86_64": [
-            "//bazel/foreign_cc:ipp-crypto",
+            ":ipp-crypto",
         ],
         "//conditions:default": [],
     }),
```

File: api/contrib/envoy/extensions/private_key_providers/cryptomb/v3alpha/cryptomb.proto
Comment (ipuustin) [reply api/contrib/envoy/extensions/private_key_providers/cryptomb/v3alpha/cryptomb.proto]: Ok, fixed.

File: docs/root/intro/arch_overview/security/ssl.rst
Comment (ipuustin) [reply docs/root/intro/arch_overview/security/ssl.rst]: Ok, I reverted the docs change. I think we can re-add the docs when the contrib doc section is there.

**Comment by ipuustin** [comment]
It looks like reverting the docs change does not work, because the CI docs precheck expects that every single protobuf message is linked to some documentation page:

```
/tmp/tmpemuw16av/generated/rst/api-v3/extensions/private_key_providers/cryptomb/v3alpha/cryptomb.proto.rst: WARNING: document isn't included in any toctree
```

So we either need to link the cryptomb protobuf message description to the root docs (by reverting the previous revert) or introduce a new "contrib" documentation tree. Or alternatively we can fix the CI so that it allows contrib extensions to have no documentation for their protobuf types. @ggreenway, what do you think?

**Comment by rojkov** [comment]
> So we either need to link the cryptomb protobuf message description to the root docs (by reverting the previous revert) or introduce a new "contrib" documentation tree. Or alternatively we can fix the CI so that it allows contrib extensions to have no documentation for their protobuf types.

Perhaps @phlax could advise us on docs in contrib

**Comment by ggreenway** [comment]
@ipuustin Thanks for your patience on this; we're still working out the kinks of `contrib`.

Can you please add a new doc section at `docs/root/api-v3/config/contrib/contrib.rst`? That file will contain a `toctree`; see other similar files conder `docs/root/api-v3/config`. Link to your new filter from there. And other contrib filters will be added to the same list.

**Comment by ggreenway** [comment]
And you'll need to link to the new `.rst` from `docs/root/api-v3/config/config.rst`.

**Comment by ggreenway** [comment]
The ARM build is broken. It took me awhile to find the actual error in the logs, but it is: 

`clang-11: error: the clang compiler does not support '-march=icelake-server'`

I don't understand why the ipp-crypto library is being built in this case, because the dependency on it is selecting for linux-x86_64, and this is ARM64 [here](https://github.com/envoyproxy/envoy/pull/17826/files#diff-bb3d9f955f48966e048dec848f898a0a7e2cee788ce7d1256473b9bb7ffc0a5aR46).

If you can't figure out why this is being built either, try to find some help on slack in the bazel or envoy-dev channels.

/wait

**Comment by ggreenway** [comment]
But the doc fix looks good; thanks for taking care of that.

**Comment by ipuustin** [comment]
> The ARM build is broken. It took me awhile to find the actual error in the logs, but it is:
> 
> `clang-11: error: the clang compiler does not support '-march=icelake-server'`
> 
> I don't understand why the ipp-crypto library is being built in this case, because the dependency on it is selecting for linux-x86_64, and this is ARM64 [here](https://github.com/envoyproxy/envoy/pull/17826/files#diff-bb3d9f955f48966e048dec848f898a0a7e2cee788ce7d1256473b9bb7ffc0a5aR46).
> 
> If you can't figure out why this is being built either, try to find some help on slack in the bazel or envoy-dev channels.
> 
> /wait

So it turned out that do_ci.sh script is building `//contrib/...` for coverage tests (see https://github.com/envoyproxy/envoy/blob/main/ci/do_ci.sh#L193). That includes all targets in the contrib tree, including ipp-crypto library. It's now "fixed" by removing ipp-crypto from `:all` target for all but `linux_x86_64` platform. A more future-proof fix might be to modify CI so that it builds only the Envoy-internal bazel targets for coverage in the contrib tree. This however would require bit more design (Would we do it by tagging targets to be outside of automatic build? Or have a denylist? Or have a /contrib/bazel/ directory for all targets not to be built as part of `:all`?) and should probably be a separate PR.

**Comment by ipuustin** [comment]
@ggreenway: Please take another look. Do you think the bazel fixes for coverage targets are now right?

**Review by ggreenway — APPROVED** [review]
This all LGTM, but I want a bazel expert to sign off on the bazel changes.

cc @lizan @wrowe. Can either of you take a look at just the bazel changes here?

**Pull request merged**
